dev/PowerNotifications/PowerNotifications.cpp

```diff
@@ -0,0 +1,240 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include <pch.h>
+#include <PowerNotifications.h>
+#include <PowerNotificationsPal.h>
+#include <PowerManager.g.cpp>
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    auto factory = make_self<factory_implementation::PowerManager>();
+
+    // EnergySaverStatus Functions
+    EventType& EnergySaverStatus_Event()
+    {
+        return factory->m_EnergySaverStatusChangedEvent;
+    }
+
+    void EnergySaverStatus_Register()
+    {
+        check_hresult(RegisterEnergySaverStatusChangedListener(&PowerManager::EnergySaverStatusChanged_Callback, &factory->m_EnergySaverStatusHandle));
+    }
+
+    void EnergySaverStatus_Unregister()
+    {
+        check_hresult(UnregisterEnergySaverStatusChangedListener(factory->m_EnergySaverStatusHandle));
+    }
+
+    void EnergySaverStatus_Update()
+    {
+        check_hresult(GetEnergySaverStatus(&factory->m_cachedEnergySaverStatus));
+    }
+
+    // BatteryStatus Functions
+    EventType& BatteryStatus_Event()
+    {
+        return factory->m_batteryStatusChangedEvent;
+    }
+
+    void BatteryStatus_Register()
+    {
+        check_hresult(RegisterCompositeBatteryStatusChangedListener(&PowerManager::CompositeBatteryStatusChanged_Callback, &factory->m_batteryStatusHandle));
+    }
+
+    void BatteryStatus_Unregister()
+    {
+        check_hresult(UnregisterCompositeBatteryStatusChangedListener(factory->m_batteryStatusHandle));
+    }
+
+    void BatteryStatus_Update()
+    {
+        check_hresult(GetCompositeBatteryStatus(&factory->m_cachedCompositeBatteryStatus));
+        factory->ProcessCompositeBatteryStatus(*factory->m_cachedCompositeBatteryStatus);
+    }
+   
+    // PowerSupplyStatus Functions
+    EventType& PowerSupplyStatus_Event()
+    {
+        return factory->m_powerSupplyStatusChangedEvent;
+    }
+
+    void PowerSupplyStatus_Register()
+    {
+        BatteryStatus_Register();
```

Comment:
This means that if an app registers both for battery status and power supply status, we double-register for composite battery changes (and leak a registration). And then when they unregister from one, we unregister the composite battery change even though the other event is still registered by the app. And then when the app does the second unregister, we unregister an invalid handle.

Multiple events fired from a single registration makes this model a little trickier to manage since the relationship between events and registrations is not 1-1. (The original code had this same problem.)